### PR TITLE
Fixed #31509 -- Made DiscoverRunner enable faulthandler by default.

### PR DESCRIPTION
**Pull request by ovkulkarni**
I mention some issues with this here: https://code.djangoproject.com/ticket/31509#comment:5

### CODE REVIEW

File: django/test/runner.py [hunk 1]
```diff
@@ -507,12 +510,16 @@ def add_arguments(cls, parser):
         )
         parser.add_argument(
             '--pdb', action='store_true',
-            help='Runs a debugger (pdb, or ipdb if installed) on error or failure.'
+            help='Runs a debugger (pdb, or ipdb if installed) on error or failure.',
```
Comment (adamchainz) [inline django/test/runner.py:513]: undo unrelated change here (revert the comma)

File: django/test/runner.py [hunk 2]
```diff
@@ -448,6 +449,8 @@ def __init__(self, pattern=None, top_level=None, verbosity=1,
         self.parallel = parallel
         self.tags = set(tags or [])
         self.exclude_tags = set(exclude_tags or [])
+        if not no_faulthandler:
+            faulthandler.enable()
```
Comment (adamchainz) [inline django/test/runner.py:453]: You need to enable faulthandler in the `ParallelTestSuite`'s processes as well, as per comment on ticket.

Comment (ovkulkarni) [reply django/test/runner.py:453]: The `faulthandler.enable()` installs signal handlers which are inherited by the child processes being forked by `multiprocessing`, so the traceback is printed even when tests are being run in parallel and the child processes aborts.

**Comment by ovkulkarni** [comment]
Sorry about that. I will write documentation and tests for this.

File: django/test/runner.py (hunk 2 above)
Comment (adamchainz) [reply django/test/runner.py:453]: Ah right. You noted on the ticket that's not the case with the 'spawn' mode, but I guess we shouldn't concern ourselves with that right now, as the test runner doesn't support that.

@Valze - maybe make a note for your spawn-mode GSoC project? 😉

**Comment by adamchainz** [comment]
> Sorry about that. I will write documentation and tests for this.

No worries.

Great work so far, thank you.

File: django/test/runner.py (hunk 2 above)
Comment (ovkulkarni) [reply django/test/runner.py:453]: Adding support for the 'spawn' mode might be pretty useful as macOS has switched to 'spawn' as the default starting at Python 3.8

**Comment by ovkulkarni** [comment]
Quick question - where is the appropriate place to write documentation?

**Comment by adamchainz** [comment]
Please read the appropiate documentation on writing documentation linked from https://docs.djangoproject.com/en/dev/internals/contributing/ : https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/ .

The new argument to the test runner should be documented alongside the others like ``--parallel`` - just grep for where that is.

The release note should be in `3.1.txt`, but since the cutoff for features is monday, you might need to rebase and target 3.2.

File: docs/ref/django-admin.txt [hunk 3]
```diff
@@ -1513,6 +1513,13 @@ installed, ``ipdb`` is used instead.
 Discards output (``stdout`` and ``stderr``) for passing tests, in the same way
 as :option:`unittest's --buffer option<unittest.-b>`.
 
+.. django-admin-option:: --no-faulthandler
+
+.. versionadded:: 3.1
+
+Disables the :py:mod:`faulthandler` module, which will print tracebacks
+if a test aborts unexpectedly, during testing.
```
Comment (adamchainz) [inline docs/ref/django-admin.txt:1521]: I suggest rearranging to first document the feature, then explain the flag. Also we can point at `faulthandler.enable()` directly, to point out a bit that it's not using faulthandler timeouts.

```suggestion
Django automatically calls :func:`faulthandler.enable()` at the start of tests,
which allows it to print a traceback if the interpreter crashes. Pass
``--no-faulthandler`` to disable this behaviour.
```

File: tests/test_runner/test_discover_runner.py [hunk 4]
```diff
@@ -297,6 +298,16 @@ def test_buffer_mode_test_fail(self):
...
 
+    @patch('faulthandler.enable')
+    def test_faulthandler_enabled(self, mocked_faulthandler):
+        _ = DiscoverRunner(no_faulthandler=False)
+        mocked_faulthandler.assert_called()
+
+    @patch('faulthandler.enable')
+    def test_faulthandler_disabled(self, mocked_faulthandler):
+        _ = DiscoverRunner(no_faulthandler=True)
+        mocked_faulthandler.assert_not_called()
```
Comment (adamchainz) [inline tests/test_runner/test_discover_runner.py:309]: 1. It's best practice to name mocks after the function
2. You don't need `_ =`. Python discards unused return values by default.

```suggestion
    @patch('faulthandler.enable')
    def test_faulthandler_enabled(self, mocked_enable):
        DiscoverRunner(no_faulthandler=False)
        mocked_enable.assert_called()

    @patch('faulthandler.enable')
    def test_faulthandler_disabled(self, mocked_enable):
        DiscoverRunner(no_faulthandler=True)
        mocked_enable.assert_not_called()
```

**Comment by adamchainz** [comment]
I am not sure offhand why the tests are failing.

File: django/test/runner.py [hunk 5]
```diff
@@ -510,7 +510,7 @@ def add_arguments(cls, parser):
         )
         parser.add_argument(
             '--pdb', action='store_true',
-            help='Runs a debugger (pdb, or ipdb if installed) on error or failure.',
+            help='Runs a debugger (pdb, or ipdb if installed) on error or failure.'
```
[the remark this answers is not in the document]
Comment (ovkulkarni) [reply django/test/runner.py]: This is a revert of the previous commit, where I added the comma. I think if you look at the total PR diff this will not show up.

[the remark this answers is not in the document]
Comment (adamchainz) [reply django/test/runner.py]: Good point :)

**Comment by ovkulkarni** [comment]
I believe the tests are failing because Jenkins is overwriting `sys.stderr` and `faulthandler.enable()` tries to call `fileno` on it (which isn't supported).

**Comment by adamchainz** [comment]
Indeed:

```
In [1]: import faulthandler

In [2]: import sys

In [3]: import io

In [4]: sys.stderr = io.StringIO()

In [5]: faulthandler.enable()
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
<ipython-input-9-7989c6ca1bf2> in <module>
----> 1 faulthandler.enable()

UnsupportedOperation: fileno
```

I can't reproduce it locally. I'm not sure how it's replaced, it would be part of the Python process.

**Comment by ovkulkarni** [comment]
We could always use `faulthandler.enable(file=sys.__stderr__)`, which according to https://docs.python.org/3/library/sys.html#sys.__stderr__ will point to the original `sys.stderr` when the python program was started.

**Comment by ovkulkarni** [comment]
Should I target this at 3.2 now?

File: django/test/runner.py (hunk 2 above)
Comment (Valze) [reply django/test/runner.py:453]: @adamchainz Sorry for the late reply, didn't catch this in my emails. 

Definitely noting this and adding it to my forum post. I'm taking my time going through my original PR and trying to flesh out the PR more and noting the current blockers.

Thanks for the catch.

**Comment by adamchainz** [comment]
> We could always use `faulthandler.enable(file=sys.__stderr__)`

That seems reasonable to me. But I'd like to understand how Jenkins is overriding `stderr`.

Another workaround is `from sys import stderr`, which will grab what stderr points to at import time, even if something else overrides `sys.stderr`. I'm not sure that's better than `sys.__stderr__` though, since it will only work if `stderr == __stderr__` anyway.

> Should I target this at 3.2 now?

Yes please.

File: django/test/runner.py [hunk 6]
```diff
@@ -434,7 +436,7 @@ def __init__(self, pattern=None, top_level=None, verbosity=1,
                  interactive=True, failfast=False, keepdb=False,
                  reverse=False, debug_mode=False, debug_sql=False, parallel=0,
                  tags=None, exclude_tags=None, test_name_patterns=None,
-                 pdb=False, buffer=False, **kwargs):
+                 pdb=False, buffer=False, no_faulthandler=False, **kwargs):
```
Comment (cjerdonek) [inline django/test/runner.py:439]: I would suggest changing the code so you're able to use the simpler `faulthandler=True` as the argument name. This will also simplify the code below that reads `if not no_faulthandler`. (You can still keep `--no-faulthandler` as the exposed option though.)

Comment (felixxm) [reply django/test/runner.py:439]: Agreed.

Comment (adamchainz) [reply django/test/runner.py:439]: I don't disagree.

Comment (ovkulkarni) [reply django/test/runner.py]: using `faulthandler=True` puts the `faulthandler` module out of scope. Should I change it to something else, or keep it as `no_faulthandler`?

File: django/test/runner.py [hunk 7]
```diff
@@ -448,6 +450,8 @@ def __init__(self, pattern=None, top_level=None, verbosity=1,
         self.parallel = parallel
         self.tags = set(tags or [])
         self.exclude_tags = set(exclude_tags or [])
+        if not no_faulthandler:
+            faulthandler.enable(file=sys.__stderr__)
```
Comment (felixxm) [inline django/test/runner.py:454]: `file` supports file descriptor so we can use the same workaround like `pytest`:
```python
try:
    faulthandler.enable(file=sys.stderr.fileno())
except (AttributeError, io.UnsupportedOperation):
    faulthandler.enable(file=sys.__stderr__.fileno())
```

File: django/test/runner.py [hunk 8]
```diff
@@ -448,6 +450,8 @@ def __init__(self, pattern=None, top_level=None, verbosity=1,
         self.parallel = parallel
         self.tags = set(tags or [])
         self.exclude_tags = set(exclude_tags or [])
+        if not no_faulthandler:
```
Comment (felixxm) [inline django/test/runner.py:453]: I think we should check if it's not already enabled:
```python
if faulthandler.is_enabled() and enable_faulthandler:
```

File: django/test/runner.py (hunk 6 above)
Comment (felixxm) [reply django/test/runner.py:439]: `no_faulthandler=False` -> `enable_faulthandler=True`

File: docs/releases/3.2.txt [hunk 9]
```diff
@@ -258,6 +258,11 @@ Tests
   now isolated for each test method. Such objects are now required to support
   creating deep copies with :py:func:`copy.deepcopy`. Assigning objects which
   don't support ``deepcopy()`` is deprecated and will be removed in Django 4.1.
+* :class:`~django.test.runner.DiscoverRunner` now enables
+  :py:mod:`faulthandler` by default. This can be disabled using
+  the :option:`test --no-faulthandler` option.
+
+* ...
```
Comment (felixxm) [inline docs/releases/3.2.txt:265]: Chop these lines.

**Comment by felixxm** [comment]
@ovkulkarni Thanks :+1: I pushed edits, added missing docs changes, added test for already enabled `faulthandler` and for `sys.stderr` that is not a file.

I also doc'd arguments added in Django 3.0 and 3.1.

**Comment by felixxm** [comment]
@adamchainz Do you want to take a look at my changes?
